dandiapi/api/tests/test_permission.py

```diff
@@ -0,0 +1,117 @@
+        ('put', '/api/dandisets/{dandiset.identifier}/users/', False),
+        # Versions
+        ('get', '/api/dandisets/{dandiset.identifier}/versions/', False),
+        ('get', '/api/dandisets/{dandiset.identifier}/versions/draft/', False),
+        ('put', '/api/dandisets/{dandiset.identifier}/versions/draft/', True),
+        ('delete', '/api/dandisets/{dandiset.identifier}/versions/draft/', True),
+        ('get', '/api/dandisets/{dandiset.identifier}/versions/draft/info/', False),
+        ('post', '/api/dandisets/{dandiset.identifier}/versions/draft/publish/', True),
+        # Assets
+        ('get', '/api/assets/{asset.asset_id}/', False),
+        ('get', '/api/assets/{asset.asset_id}/download/', False),
+        ('get', '/api/assets/{asset.asset_id}/info/', False),
+        ('get', '/api/dandisets/{dandiset.identifier}/versions/draft/assets/', False),
+        ('post', '/api/dandisets/{dandiset.identifier}/versions/draft/assets/', True),
+        ('get', '/api/dandisets/{dandiset.identifier}/versions/draft/assets/paths/', False),
+        (
+            'get',
+            '/api/dandisets/{dandiset.identifier}/versions/draft/assets/{asset.asset_id}/',
+            False,
+        ),
+        (
+            'put',
+            '/api/dandisets/{dandiset.identifier}/versions/draft/assets/{asset.asset_id}/',
+            True,
+        ),
+        (
+            'delete',
+            '/api/dandisets/{dandiset.identifier}/versions/draft/assets/{asset.asset_id}/',
+            True,
+        ),
+        (
+            'get',
+            '/api/dandisets/{dandiset.identifier}/versions/draft/assets/{asset.asset_id}/download/',
+            False,
+        ),
+        (
+            'get',
+            '/api/dandisets/{dandiset.identifier}/versions/draft/assets/{asset.asset_id}/info/',
+            False,
+        ),
+        (
+            'get',
+            '/api/dandisets/{dandiset.identifier}/versions/draft/assets/{asset.asset_id}/validation/',  # noqa: E501
+            False,
+        ),
+        # Zarrs
+        ('get', '/api/zarr/', False),
+        ('post', '/api/zarr/', True),
+        ('get', '/api/zarr/{zarr.zarr_id}/', False),
+        ('delete', '/api/zarr/{zarr.zarr_id}/files/', True),
+        ('post', '/api/zarr/{zarr.zarr_id}/upload/', True),
+        ('delete', '/api/zarr/{zarr.zarr_id}/upload/', True),
+        ('post', '/api/zarr/{zarr.zarr_id}/upload/complete/', True),
+    ],
+)
+@pytest.mark.django_db
+def test_approved_or_readonly(
+    api_client,
+    user,
+    dandiset_factory,
+    draft_version_factory,
+    draft_asset_factory,
+    zarr_archive_factory,
+    method,
+    url_format,
+    owner_required,
+):
+    dandiset = dandiset_factory()
+    version = draft_version_factory(dandiset=dandiset)
+    zarr = zarr_archive_factory(dandiset=dandiset)
+    asset = draft_asset_factory()
+    version.assets.add(asset)
+
+    url = url_format.format(dandiset=dandiset, asset=asset, zarr=zarr)
```

Comment:
This is cool, for some reason it never occurred to me that you can do this to pass data from other fixtures into parameterized strings :open_mouth:

Reply:
Credits to @dchiquito for originally doing this in the embargo visibility tests.